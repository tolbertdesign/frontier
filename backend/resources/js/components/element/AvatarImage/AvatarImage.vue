<template>
  <img
    :alt="alt"
    :src="src"
    :class="[size, 'shadow border rounded-full']"
    @click="emit"
  >
</template>

<script>
export default {
  name: 'AvatarImage',
  status: 'prototype',
  release: '1.0.0',
  props: {
    alt: {
      type: String,
      default: '',
    },
    src: {
      type: String,
      default: '/v3-assets/dashboard/images/dashboard-avatar.svg',
    },
    shouldEmit: {
      type: Boolean,
      default: false,
    },
    small: {
      type: Boolean,
      default: false,
    },
  },
  computed: {
    size () {
      return this.small ? 'w-9 h-9' : 'w-16 h-16'
    },
  },
  methods: {
    emit () {
      if (this.shouldEmit) {
        this.$emit('open')
      }
    },
  },
}
</script>
