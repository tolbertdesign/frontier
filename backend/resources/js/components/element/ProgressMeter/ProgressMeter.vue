<template>
  <div :class="['meter', backgroundColorClass, (showValue && firstMeterPercentage === 0 ? 'noValueDisplay pb-6': '')]">
    <span
      :style="{ width: firstMeterPercentage + '%' }"
      :class="['meter-1', color === 'secondary' ? 'has-background-secondary': 'bg-primary' ]"
    />
    <span
      v-if="showValue"
      :style="getPercentagePosition(firstMeterPercentage)"
      class="valueDisplay font-semibold"
    >
      {{ valueToDisplay }}
    </span>
    <span
      :style="{ width: secondMeterPercentage + '%' }"
      class="meter-2"
    />
  </div>
</template>

<script>
export default {
  name: 'ProgressMeter',
  status: 'prototype',
  release: '1.0.0',
  props: {
    color: {
      type: String,
      default: 'primary',
    },
    backgroundColorClass: {
      type: String,
      default: 'bg-grey-light',
    },
    goal: {
      type: [String, Number],
      default: 100,
    },
    firstMeter: {
      type: Number,
      default: 0,
    },
    secondMeter: {
      type: Number,
      default: 0,
    },
    showValue: {
      type: Boolean,
      default: false,
    },
    valueToDisplay: {
      type: String,
      default: '',
    },
  },
  computed: {
    firstMeterPercentage () {
      return this.getPercentage(this.firstMeter, this.goal)
    },
    secondMeterPercentage () {
      return this.getPercentage(this.firstMeter + this.secondMeter, this.goal)
    },
  },
  methods: {
    getPercentage (part, total) {
      if (isNaN(part) || isNaN(total) || total <= 0) {
        return 0
      }
      const percent = part / total * 100
      if (percent > 100) {
        return 100
      }
      return percent
    },
    getPercentagePosition (percent) {
      if (percent > 15) {
        return { right: Math.abs((percent - 100)) + '%' }
      } else {
        return { left: '0%' }
      }
    },
  },
}
</script>
