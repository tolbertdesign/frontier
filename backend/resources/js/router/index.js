import Vue from 'vue'
import VueRouter from 'vue-router'
import AppContainer from '@/layouts/AppContainer'
import NProgress from 'nprogress'
import ParentDashboard from '@/views/ParentDashboard'
import TeacherDashboard from '@/views/TeacherDashboard'
import EditParticipant from '@/views/EditParticipant'
import EditProfile from '@/views/EditProfile'
import EasyEmailer from '@/views/EasyEmailer/EasyEmailer'
import FinishLine from '@/views/FinishLine'
import store from '@/store'

Vue.use(VueRouter)

const routes = [
  {
    path: '/v3/home',
    component: AppContainer,
    children: [
      {
        path: 'dashboard',
        name: 'parent-dashboard',
        component: ParentDashboard,
      },
      {
        path: 'teacher-dashboard',
        name: 'teacher-dashboard',
        component: TeacherDashboard,
      },
      {
        path: 'edit-participant/:id',
        name: 'edit-participant',
        component: EditParticipant,
      },
      {
        path: 'edit-profile',
        name: 'edit-profile',
        component: EditProfile,
      },
      {
        path: 'easy-emailer/:participantUserId',
        name: 'easy-emailer',
        component: EasyEmailer,
      },
      {
        path: 'finish-line/:participantUserId',
        name: 'finish-line',
        component: FinishLine,
      },
    ],
  },

]

const router = new VueRouter({
  mode: 'history',
  forcePageRouteRefresh: false,
  linkActiveClass: 'is-active',
  routes,

  scrollBehavior (to, from, savedPosition) {
    return { x: 0, y: 0 }
  },
})

router.beforeEach((routeTo, routeFrom, next) => {
  NProgress.start()
  store.dispatch('setContentGroup', routeTo)

  if (router.options.forcePageRouteRefresh) {
    window.location.href = '/v3/home/dashboard/' + routeTo.path
    router.options.forcePageRouteRefresh = false
  } else {
    next()
  }
})

router.afterEach(() => {
  NProgress.done()
})

export default router
