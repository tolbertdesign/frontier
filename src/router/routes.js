import store from '@/vuex/store'

export default [
  {
    path: '/',
    name: 'welcome',
    component: () => import(/* webpackChunkName: "welcome" */ '@/pages/welcome'),
    meta: {
      layout: 'FullScreen',
      requiresVisitor: true,
    },
  },
  {
    path: '/login',
    name: 'login',
    component: () => import(/* webpackChunkName: "login" */ '@/pages/login'),
    meta: {
      layout: 'FullScreen',
      requiresVisitor: true,
    },
  },
  {
    path: '/register',
    name: 'register',
    component: () => import(/* webpackChunkName: "register" */ '@/pages/register'),
    meta: {
      layout: 'FullScreen',
      requiresVisitor: true,
    },
  },
  {
    path: '/buy',
    name: 'buy',
    component: () => import(/* webpackChunkName: "buy" */ '@/pages/buy'),
    meta: {
      layout: 'FlexboxApp',
    },
  },
  {
    path: '/contact',
    name: 'contact',
    component: () => import(/* webpackChunkName: "contact" */ '@/pages/contact'),
    meta: {
      requiresAuth: true,
    },
  },
  {
    path: '/dashboard',
    name: 'dashboard',
    component: () => import(/* webpackChunkName: "dashboard" */ '@/pages/dashboard'),
    meta: {
      requiresAuth: true,
    },
  },
  {
    path: '/landing',
    name: 'landing',
    component: () => import(/* webpackChunkName: "landing" */ '@/pages/landing'),
    meta: {
      requiresAuth: true,
    },
  },
  {
    path: '/templates',
    name: 'templates',
    component: () => import(/* webpackChunkName: "templates" */ '@/pages/templates'),
    meta: {
      layout: 'Sidebar',
      requiresAuth: true,
    },
  },
  {
    path: '/documents',
    name: 'documents',
    component: () => import(/* webpackChunkName: "documents" */ '@/pages/documents'),
    meta: {
      layout: 'Sidebar',
      requiresAuth: true,
    },
  },
  {
    path: '/reports',
    name: 'reports',
    component: () => import(/* webpackChunkName: "reports" */ '@/pages/reports'),
    meta: {
      layout: 'Sidebar',
      requiresAuth: true,
    },
  },
  {
    path: '/profile',
    name: 'profile',
    component: () => import(/* webpackChunkName: "edit-profile" */ '@/pages/profile'),
    meta: {
      requiresAuth: true,
    },
  },
  {
    path: '/settings',
    name: 'settings',
    component: () => import(/* webpackChunkName: "settings" */ '@/pages/settings'),
    meta: {
      requiresAuth: true,
    },
  },
  {
    path: '/event/list',
    name: 'event-list',
    component: () => import(/* webpackChunkName: "event-list" */ '@/pages/event-list'),
    props: true,
  },
  {
    path: '/event/create',
    name: 'event-create',
    component: () => import(/* webpackChunkName: "event-create" */ '@/pages/event-create'),
  },
  {
    path: '/event/:id',
    name: 'event-show',
    component: () => import(/* webpackChunkName: "event-show" */ '@/pages/event-show'),
    props: true,
    beforeEnter (routeTo, routeFrom, next) {
      store
        .dispatch('event/fetchEvent', routeTo.params.id)
        .then(event => {
          routeTo.params.event = event
          next()
        })
        .catch(error => {
          if (error.response && error.response.status == 404) {
            next({ name: '404', params: { resource: 'event' } })
          } else {
            next({ name: 'network-issue' })
          }
        })
    },
  },
  {
    path: '/user/:id',
    component: () => import(/* webpackChunkName: "user" */ '@/pages/user'),

  },
  {
    path: '/404',
    name: '404',
    component: () => import(/* webpackChunkName: "not-found" */ '@/pages/not-found'),
  },
  {
    path: '/network-issue',
    name: 'network-issue',
    component: () => import(/* webpackChunkName: "network-issue" */ '@/pages/network-issue'),
  },
  {
    path: '*',
    redirect: {
      name: '404',
      params: {
        resource: 'page',
      },
    },
  },
]
