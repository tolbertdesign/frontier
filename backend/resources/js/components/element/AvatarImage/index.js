import AvatarImage from './AvatarImage.vue'

export default AvatarImage
