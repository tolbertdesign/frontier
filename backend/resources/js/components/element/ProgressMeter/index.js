import ProgressMeter from './ProgressMeter.vue'

export default ProgressMeter
