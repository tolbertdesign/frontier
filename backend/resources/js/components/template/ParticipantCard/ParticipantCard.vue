<template>
  <div class="participant-card max-w-xl m-4">
    <div>
      <div class="media mb-2 lg:px-4 relative">
        <div class="media-left">
          <a @click="editParticipant(participant)">
            <AvatarImage
              :alt="participant.first_name"
              :src="avatarSrc"
            />
          </a>
        </div>
        <div class="media-content self-center">
          <h2 class="text-xl">
            {{ participant.first_name }}
          </h2>
        </div>
        <div class="media-right self-start has-text-right">
          <a
            class="text-sm has-text-grey"
            @click="editParticipant(participant)"
          >
            {{ lang.participant_card.edit_profile_link }}
          </a>
        </div>
      </div>
    </div>
    <section class="lg:px-4">
      <div
        class="mb-4"
        style="display: block"
      >
        <ProgressMeter
          :first-meter="totalPledged"
          :second-meter="totalPending"
          :goal="participant.profile.pledge_goal"
        />

        <div class="mt-2">
          <h3 class="font-bold flex items-center">
            <svg
              class="fill-current text-primary h-4 w-4 mr-1"
              viewBox="0 0 16 16"
            >
              <circle
                cx="8"
                cy="8"
                r="8"
              />
            </svg>
            <span>
              {{ lang.participant_card.total_pledged }} ${{ totalPledged | currency }} {{ lang.participant_card.of }} ${{ participant.profile.pledge_goal | currency }}
            </span>
          </h3>
          <h4
            v-if="!isFlat"
            class="text-sm capitalize"
          >
            {{ program.unit.modifier }} {{ program.unit.name }}
          </h4>
        </div>
        <div
          v-if="totalPending > 0"
          class="mt-2"
        >
          <h3 class="font-bold flex items-center">
            <svg
              class="fill-current text-tertiary h-4 w-4 mr-1"
              viewBox="0 0 16 16"
            >
              <circle
                cx="8"
                cy="8"
                r="8"
              />
            </svg>
            <span>
              {{ lang.participant_card.awaiting_confirmation }} ${{ totalPending | currency }}
            </span>
          </h3>
          <h4
            v-if="!isFlat"
            class="text-sm capitalize"
          >
            {{ program.unit.modifier }} {{ program.unit.name }}
          </h4>
        </div>
      </div>
      <div
        :class="[ selected == participant.id ? '' : 'border-b lg:border-b-0' ]"
        class="button-container text-center pb-4"
      >
        <button
          class="py-2 px-8 rounded-full font-semibold text-sm text-white bg-secondary shadow focus:outline-none focus:shadow-outline"
          @click="$emit('toggle', participant.id)"
        >
          <span>{{ selected == participant.id ? lang.close : lang.view }} {{ lang.rewards }}</span>
        </button>
      </div>
    </section>
  </div>
</template>

<script>
import AvatarImage from '@/components/element/AvatarImage'
import ProgressMeter from '@/components/element/ProgressMeter'
import PledgeTotalCalculator from '@/utilities/PledgeTotalCalculator.js'

export default {
  name: 'ParticipantCard',
  status: 'prototype',
  release: '1.0.0',
  components: {
    AvatarImage,
    ProgressMeter,
  },
  props: {
    program: {
      type: Object,
      default: () => ({
        program_pledge_setting: {
          flat_donate_only: 0,
        },
        unit: {
          per: null,
          modifier: null,
        },
        unit_flat_conversion: null,
      }),
    },
    participant: {
      type: Object,
      default: null,
    },
    selected: {
      type: Number,
      default: 0,
    },
    group: {
      type: Number,
      default: 0,
    },
  },

  data () {
    return {
      pledgeStatuses: [
        {
          name: 'confirmed',
          id: 2,
        },
        {
          name: 'paid',
          id: 3,
        },
        {
          name: 'pending',
          id: 4,
        },
        {
          name: 'paymentScheduled',
          id: 8,
        },
      ],
    }
  },

  computed: {
    lang () {
      return this.$store.state.lang
    },
    avatarSrc () {
      return this.participant.profile.image_name == null ? '/v3-assets/dashboard/images/dashboard-avatar.svg' : this.avatarImgSrc
    },
    avatarImgSrc () {
      return this.$store.getters.avatarPath + this.participant.profile.image_name
    },
    editProfileLink () {
      return '/v3/tkdashboard/?redirect=' + encodeURI('auth/login/' + this.participant.fr_code + '/edit-personalize')
    },
    isFlat () {
      return this.program.program_pledge_setting.flat_donate_only === 1
    },
    totalPending () {
      const statusNames = ['pending']
      const statusIds = this.pledgeStatuses
        .filter(status => statusNames.includes(status.name))
        .map(status => status.id)
      return this.getPledgeTotal(statusIds)
    },
    totalPledged () {
      const statusNames = ['confirmed', 'paid', 'paymentScheduled']
      const statusIds = this.pledgeStatuses
        .filter(status => statusNames.includes(status.name))
        .map(status => status.id)
      return parseFloat(this.getPledgeTotal(statusIds))
    },
  },

  methods: {
    getPledgeTotal (statusIds) {
      if (this.program.unit_flat_conversion <= 0) {
        return 0
      }
      if (this.isFlat) {
        return this.participant.participant_info.pledges
          .filter(pledge => statusIds.includes(pledge.pledge_status_id))
          .reduce((total, pledge) => total + pledge.amount, 0)
      }
      const calculator = new PledgeTotalCalculator(this.program, this.participant.participant_info.pledges)
      return calculator.getTotalAsPPL()
    },
    editParticipant (participant) {
      this.$router.push({
        name: 'edit-participant',
        params: {
          id: participant.id,
        },
      })
    },
  },
}
</script>
