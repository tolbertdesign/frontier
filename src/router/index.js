import Vue from 'vue'
import Router from 'vue-router'
// Adds a loading bar at the top during page loads.
import NProgress from 'nprogress'
import 'nprogress/nprogress.css'

// https://github.com/declandewet/vue-meta
import Meta from 'vue-meta'
import { sync } from 'vuex-router-sync'
import routes from '@/router/routes'
import store from '@/vuex/store'

Vue.use(Router)
Vue.use(Meta, {
  // The component option name that vue-meta looks for meta info on.
  keyName: 'page',
})

const router = new Router({
  routes,
  base: process.env.BASE_URL,
  // Use the HTML5 history API (i.e. normal-looking routes)
  // instead of routes with hashes (e.g. example.com/#/about).
  // This may require some server configuration in production:
  // https://router.vuejs.org/en/essentials/history-mode.html#example-server-configurations
  mode: 'history',
  // Simulate native-like scroll behavior when navigating to a new
  // route and using back/forward buttons.
  scrollBehavior (to, from, savedPosition) {
    if (savedPosition) {
      return savedPosition
    } else {
      return { x: 0, y: 0 }
    }
  },
})

router.beforeEach((routeTo, routeFrom, next) => {
  // If this isn't an initial page load...
  if (routeFrom.name !== null) {
    // Start the route progress bar.
    NProgress.start()
  }

  if (routeTo.matched.some(record => record.meta.requiresAuth)) {
    if (!isLoggedIn()) {
      next({
        name: 'home',
      })
    } else {
      next()
    }
  } else if (routeTo.matched.some(record => record.meta.requiresVisitor)) {
    if (isLoggedIn()) {
      next({
        name: 'dashboard',
      })
    } else {
      next()
    }
  } else {
    next() // make sure to always call next()!
  }

  // If auth is required and the user is NOT currently logged in,
  // redirect to login.
  // redirectToLogin()

  // function redirectToLogin () {
  //   // Pass the original route to the login component
  //   next({ name: 'login', query: { redirectFrom: routeTo.fullPath } })
  // }
})

router.afterEach(() => {
  NProgress.done()
})

function isLoggedIn () {
  return store.getters['auth/loggedIn']
}

sync(store, router)

export default router
