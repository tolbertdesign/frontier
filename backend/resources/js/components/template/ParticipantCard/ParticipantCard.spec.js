import { shallowMount, createLocalVue } from '@vue/test-utils'
import ParticipantCard from './ParticipantCard.vue'
import router from '@/router'
import AvatarImage from '@/components/element/AvatarImage'
import ProgressMeter from '@/components/element/ProgressMeter'

const localVue = createLocalVue()
localVue.filter('currency', jest.fn())

const $store = {
  state: {
    lang: {
      participant_card: {
        edit_profile_link: 'edit profile link',
        total_pledged: 'total pledged',
        of: 'of',
        pledge_goal: 'pledge goal',
        awaiting_confirmation: 'awaiting confirmation',
      },
      close: 'close',
      view: 'view',
      rewards: 'rewards',
    },
  },
  getters: {
    avatarPath: '',
  },
}
const selected = 1
const group = 5
const pledge1 = {
  id: 11,
  pledge_status_id: 2,
  pledge_type_id: 3,
  amount: 20,
}
const pledge2 = {
  id: 22,
  pledge_status_id: 1,
  pledge_type_id: 1,
  amount: 1.50,
}
const participant = {
  id: 7,
  first_name: 'first name',
  last_name: 'last name',
  profile: {
    pledge_goal: 1000,
    image_name: 'image name',
    fr_code: 'fr code',
  },
  participant_info: {
    pledges: [pledge1, pledge2],
  },
}
const program = {
  id: 8,
  participants: [participant],
  program_pledge_setting: {
    flat_donate_only: 0,
  },
  unit: {
    per: 'per',
    modifier: 'modifier',
  },
  unit_flat_conversion: 30,
}

describe('ParticipantCard', () => {
  it('Snapshop test for ParticipantCard', () => {
    const wrapper = shallowMount(ParticipantCard, {
      localVue,
      router,
      mocks: {
        $store,
      },
      components: {
        AvatarImage,
        ProgressMeter,
      },
      propsData: {
        program: program,
        participant: participant,
        selected: selected,
        group: group,
      },
    })
    expect(wrapper).toMatchSnapshot()
  })
})
