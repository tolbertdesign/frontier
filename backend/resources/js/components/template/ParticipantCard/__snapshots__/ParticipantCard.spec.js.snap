// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ParticipantCard Snapshop test for ParticipantCard 1`] = `
<div class="participant-card max-w-xl m-4">
  <div>
    <div class="media mb-2 lg:px-4 relative">
      <div class="media-left"><a>
          <avatarimage-stub alt="first name" src="image name"></avatarimage-stub>
        </a></div>
      <div class="media-content self-center">
        <h2 class="text-xl">
          first name
        </h2>
      </div>
      <div class="media-right self-start has-text-right"><a class="text-sm has-text-grey">
          edit profile link
        </a></div>
    </div>
  </div>
  <section class="lg:px-4">
    <div class="mb-4" style="display: block;">
      <progressmeter-stub color="primary" backgroundcolorclass="bg-grey-light" goal="1000" firstmeter="0.6666666666666666" secondmeter="0.6666666666666666" valuetodisplay=""></progressmeter-stub>
      <div class="mt-2">
        <h3 class="font-bold flex items-center"><svg viewBox="0 0 16 16" class="fill-current text-primary h-4 w-4 mr-1">
            <circle cx="8" cy="8" r="8"></circle>
          </svg> <span>
                        total pledged $ of $
                    </span></h3>
        <h4 class="text-sm capitalize">
          modifier
        </h4>
      </div>
      <div class="mt-2">
        <h3 class="font-bold flex items-center"><svg viewBox="0 0 16 16" class="fill-current text-tertiary h-4 w-4 mr-1">
            <circle cx="8" cy="8" r="8"></circle>
          </svg> <span>
                        awaiting confirmation $
                    </span></h3>
        <h4 class="text-sm capitalize">
          modifier
        </h4>
      </div>
    </div>
    <div class="button-container text-center pb-4 border-b lg:border-b-0"><button class="py-2 px-8 rounded-full font-semibold text-sm text-white bg-secondary shadow focus:outline-none focus:shadow-outline"><span>view rewards</span></button></div>
  </section>
</div>
`;
